### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by sidepelican** [yoheimuta/BufferedLogger#6]
https://github.com/yoheimuta/BufferedLogger/blob/f9d434f8b8e5b331615066860d4a2be68fbfc5e8/BufferedLogger/BufferedOutput.swift#L170-L190

this closure is called by `writer` and the thread is depending it.

I think it need to wrap this by `queue`

https://github.com/yoheimuta/BufferedLogger/blob/f9d434f8b8e5b331615066860d4a2be68fbfc5e8/BufferedLogger/BufferedOutput.swift#L172

**Comment by yoheimuta** [comment]
@sidepelican Thank you for reporting a bug.

I can confirm that the writer callback is called from the other thread.
Second dispatchPrecondition is detecting that the condition evaluates false.

```diff
/y/W/p/s/BufferedLogger ❯❯❯ gwd
diff --git a/BufferedLogger/BufferedOutput.swift b/BufferedLogger/BufferedOutput.swift
index 1164200..708f5ad 100644
--- a/BufferedLogger/BufferedOutput.swift
+++ b/BufferedLogger/BufferedOutput.swift
@@ -167,7 +167,15 @@ final class BufferedOutput {
     }

     private func callWriteChunk(_ chunk: Chunk) {
+        if #available(iOS 10.0, *) {
+            dispatchPrecondition(condition: .onQueue(queue))
+        }
+
         writer.write(chunk) { success in
+            if #available(iOS 10.0, *) {
+                dispatchPrecondition(condition: .onQueue(self.queue))
+            }
+
             if success {
                 do {
                     try self.entryStorage.remove(chunk.entries,
diff --git a/BufferedLoggerTests/BufferedOutputTests.swift b/BufferedLoggerTests/BufferedOutputTests.swift
index 3fd126c..8eb070c 100644
--- a/BufferedLoggerTests/BufferedOutputTests.swift
+++ b/BufferedLoggerTests/BufferedOutputTests.swift
@@ -32,8 +32,11 @@ final class MockWriter: Writer {
             givenPayloads.append($0.payload)
         }

-        completion(shouldSuccess)
-        writeCallback?(calledWriteCount-1)
+        let queue = DispatchQueue(label: "com.github.yoheimuta.BufferedLogger.BufferedOutputTests")
+        queue.async {
+            completion(self.shouldSuccess)
+            self.writeCallback?(self.calledWriteCount-1)
+        }
     }
 }

diff --git a/Demo/Demo/MyWriter.swift b/Demo/Demo/MyWriter.swift
index c5fa115..3f27571 100644
--- a/Demo/Demo/MyWriter.swift
+++ b/Demo/Demo/MyWriter.swift
@@ -17,6 +17,9 @@ class MyWriter: Writer {
             print("entry is \($0)")
         }

-        completion(true)
+        let queue = DispatchQueue(label: "com.github.yoheimuta.Demo.MyWriter")
+        queue.async {
+            completion(true)
+        }
     }
 }
```

I'll fix it with next PR.

**Comment by yoheimuta** [comment]
@sidepelican I created https://github.com/yoheimuta/BufferedLogger/pull/7 to try fixing this issue.

Could you look over this PR to fix this issue?

**Comment by yoheimuta** [comment]
merged #7 and shipped a new version [0.4.1](https://github.com/yoheimuta/BufferedLogger/releases/tag/0.4.1).
